src/transactions/test/PathPaymentStrictSendTests.cpp

```diff
@@ -0,0 +1,2285 @@
+            auto expected = std::vector<ClaimOfferAtom>{
+                o1a.exchanged(10, 20), o1b.exchanged(30, 60),
+                o2.exchanged(20, 40), o3.exchanged(10, 20)};
+            REQUIRE(actual == expected);
+            // clang-format off
+            market.requireBalances(
+                {{source, {{xlm, minBalance4 - 2 * txfee}, {cur1, 0}, {cur2, 0}, {cur3, 0}, {cur4, 0}}},
+                 {mm12a, {{xlm, minBalance3 - 3 * txfee}, {cur1, 20}, {cur2, 30}, {cur3, 0}, {cur4, 0}}},
+                 {mm12b, {{xlm, minBalance3 - 3 * txfee}, {cur1, 60}, {cur2, 10}, {cur3, 0}, {cur4, 0}}},
+                 {mm23, {{xlm, minBalance3 - 3 * txfee}, {cur1, 0}, {cur2, 40}, {cur3, 0}, {cur4, 0}}},
+                 {mm34, {{xlm, minBalance3 - 3 * txfee}, {cur1, 0}, {cur2, 0}, {cur3, 20}, {cur4, 0}}},
+                 {destination, {{xlm, minBalance1 - txfee}, {cur1, 0}, {cur2, 0}, {cur3, 0}, {cur4, 10}}}});
+            // clang-format on
+        });
+    }
+
+    SECTION("uses whole best offer for second exchange")
+    {
+        auto market = TestMarket{*app};
+        auto source = root.create("source", minBalance4);
+        auto destination = root.create("destination", minBalance1);
+        auto mm12 = root.create("mm12a", minBalance3);
+        auto mm23a = root.create("mm23a", minBalance3);
+        auto mm23b = root.create("mm23b", minBalance3);
+        auto mm34 = root.create("mm34", minBalance3);
+
+        source.changeTrust(cur1, 200);
+        mm12.changeTrust(cur1, 200);
+        mm12.changeTrust(cur2, 200);
+        mm23a.changeTrust(cur2, 200);
+        mm23a.changeTrust(cur3, 200);
+        mm23b.changeTrust(cur2, 200);
+        mm23b.changeTrust(cur3, 200);
+        mm34.changeTrust(cur3, 200);
+        mm34.changeTrust(cur4, 200);
+        destination.changeTrust(cur4, 200);
+
+        gateway.pay(source, cur1, 80);
+        gateway.pay(mm12, cur2, 40);
+        gateway2.pay(mm23a, cur3, 20);
+        gateway2.pay(mm23b, cur3, 20);
+        gateway2.pay(mm34, cur4, 10);
+
+        auto o1 = market.requireChangesWithOffer({}, [&] {
+            return market.addOffer(mm12, {cur2, cur1, Price{2, 1}, 40});
+        });
+        auto o2a = market.requireChangesWithOffer({}, [&] {
+            return market.addOffer(mm23a, {cur3, cur2, Price{2, 1}, 15});
+        });
+        auto o2b = market.requireChangesWithOffer({}, [&] {
+            return market.addOffer(mm23b, {cur3, cur2, Price{2, 1}, 10});
+        });
+        auto o3 = market.requireChangesWithOffer({}, [&] {
+            return market.addOffer(mm34, {cur4, cur3, Price{2, 1}, 10});
+        });
+
+        for_versions_from(12, *app, [&] {
+            auto actual = std::vector<ClaimOfferAtom>{};
+            market.requireChanges({{o1.key, OfferState::DELETED},
+                                   {o2a.key, OfferState::DELETED},
+                                   {o2b.key, {cur3, cur2, Price{2, 1}, 5}},
+                      
```

Comment:
re: "0 trade"
is there a missing test case where we end up deleting 03c in this situation? (was supposed to "stay" but got adjusted to be deleted)